server/service/redis_policy_set/redis_policy_set.go

```diff
@@ -0,0 +1,139 @@
+package redis_policy_set
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+
+	"github.com/fleetdm/fleet/v4/server/datastore/redis"
+	"github.com/fleetdm/fleet/v4/server/fleet"
+	"github.com/fleetdm/fleet/v4/server/service"
+	redigo "github.com/gomodule/redigo/redis"
+)
+
+type redisFailingPolicySet struct {
+	pool fleet.RedisPool
+}
+
+var _ service.FailingPolicySet = (*redisFailingPolicySet)(nil)
+
+// NewFailing creates a redis policy set for failing policies.
+func NewFailing(pool fleet.RedisPool) *redisFailingPolicySet {
+	return &redisFailingPolicySet{
+		pool: pool,
+	}
+}
+
+const (
+	policySetKeyPrefix = "policies:failing:"
+	// We use this approach to avoid a SCAN command when listing policy sets.
+	policySetsSetKey = "policies:failing_sets"
+)
+
+func policySetKey(policyID uint) string {
+	return policySetKeyPrefix + strconv.Itoa(int(policyID))
+}
+
+func hostEntry(host service.PolicySetHost) string {
+	return strconv.Itoa(int(host.ID)) + "," + host.Hostname
+}
+
+func parseHostEntry(v string) (*service.PolicySetHost, error) {
+	parts := strings.SplitN(v, ",", 2)
+	if len(parts) != 2 {
+		return nil, fmt.Errorf("invalid format: %s", v)
+	}
+	id, err := strconv.Atoi(parts[0])
+	if err != nil {
+		return nil, fmt.Errorf("invalid id: %s", v)
+	}
+	return &service.PolicySetHost{
+		ID:       uint(id),
+		Hostname: parts[1],
+	}, nil
+}
+
+// ListSets lists all the policy sets.
+func (r *redisFailingPolicySet) ListSets() ([]uint, error) {
+	conn := redis.ConfigureDoer(r.pool, r.pool.Get())
+	defer conn.Close()
+
+	ids, err := redigo.Ints(conn.Do("SMEMBERS", policySetsSetKey))
```

Comment:
Not an issue necessarily, but if we expect a huge number of members, `SSCAN` might be better (to prevent locking redis for too long in an SMEMBERS call). That being said, I doubt we expect many thousands of policies (and the number of members is related to the number of policies), so this is probably the right approach.